### DISCUSSION TITLE
RFC: function handlers named as http methods + api support in app dir

### DISCUSSION THREAD

**Discussion by DanieliusDev** [vercel/next.js#44803] [Ideas]
### Describe the feature you'd like to request

With the new `app` directory, I would like to see the option to allow to handle certain http methods with an export with the name of the http method. We should still be able to use the default export as a way to handle any of the methods that dont have their own handler. It would also be useful if we could make apis with this, meaning we are not returning jsx in our handler.

## Example

```tsx
// not case sensitive??
export function GET({ request, ...props }: NextProps) {
    // Handle GET requests
    return <>Hello World!</>;
}

export function POST({ request, ...props }: NextProps) {
    // Handle POST requests
    return new Response('Hello World!');
}

// If no default handler is set, it will redirect to 404 page
export default funcion Default({ request, ...props }: NextProps) {
    // Handle requests that are not GET or POST requests
    return new Response('Method not allowed!', { status: 405 });
}
```

### Describe the solution you'd like

Whenever a request is received, run the appropriate handler, for example a `GET` request will run a `GET` handler for that route. If a handler isn't defined for the specific http method (`GET` in this case) the default handler (default exported function) will run, or redirect the request to a 404 page if a default handler is not defined.

### Describe alternatives you've considered

Instead of exporting function handlers (or allow both), allow the user to make a file for a http method, for example, to handle a `GET` request make a file named `GET.tsx`. If the http method is not defined the `page.tsx` handler will run instead, or redirect the user to the 404 page if there is no `page.tsx` for the route. All method handler filess must be UPPERCASE or have a special syntax like `#{method}.*`.

**Comment by DanieliusDev** [comment]
thoughts on this idea would be appreciated.

**Comment by DanieliusDev** [comment]
bump

**Comment by robertwbradford** [comment]
13.2 introduced this: https://beta.nextjs.org/docs/routing/route-handlers
